added bytes-based storage

siennathesane opened the pull request:
This is the initial PR for `Bytes`-based storage. I'm not 100% sure if there is an API change, since I reconfigured `Default` to accept any bitmap data type.

`serde_json` is very slow to encode `BytesBitmap` (re: ~75s) versus `bincode` (re: ~150ms), so `bincode` is used in the tests.

Resolves: #16

siennathesane committed:
added bytes-based storage

Signed-off-by: Sienna Satterwhite <sienna@linux.com>

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+use bytes::{BufMut, Bytes, BytesMut};
+use std::convert::TryInto;
+
+/// A plain, heap-allocated, `O(1)` indexed bitmap using `bytes::BytesMut` for storage.
+///
+/// This bitmap requires `O(n)` space and can be read and wrote to in `O(1)` time.
+///
+/// This type is fast for both reads and writes, trading internal complexity for speed.
```
I think you should probably spell out the benefits of the `bytes` impl over the `vec` impl - like, when would someone want to use this impl vs. the others? I think "zero copy clones" and direct serialisation are nice properties to shout about!

I'd probably drop the "trading internal complexity for speed" statement (it's fairly subjective - this is much simpler than the compressed representation!) and perhaps replace it with a description of the space complexity tradeoff made by an uncompressed bitmap:

> This type provide fast `O(1)` read and write operations, but trades `O(n)` space complexity for the additional performance.

And something about serialisation like:

> The [`BytesBitmap`] representation is suitable for persistence without the need for serialisation; the output of [`BytesBitmap::freeze()`] can be used to construct a new instance. [`Serde`] serialisation is also implemented as a conveinence to enable serialisation to various formats.

siennathesane replied:
sure, what you wrote looks good to me

domodwyer commented on README.md:
```diff
@@ -75,4 +75,8 @@ is not reusable in a different process; for serialised filters a different
 hasher should be used. By default, derived [`Hash`] implementation is not
 considered portable but a hand-wrote implementation can be.
 
+If you are using the `BytesBitmap` as your bitmap storage, it is recommended to use
+the `bincode` library due to performance reasons. In initial testing, using 
+`serde_json` was very slow to encode the bitmap.
```
I'm not sure you need any serialisation / serde usage at all once you've exposed the underlying byte store with the `bloom.bitmap()` accessor - you can simply serialise it as-is! Something like:

```rust
let bloom = ...;

bloom.insert("bananas");
let data = bloom.bitmap();

save(&data); // Write the bytes to disk / network
```

You'd then reconstruct the bloom filter like this:

```rust
let data = load(); // Read the bytes from above
let bitmap = BytesBitmap::from_bytes(data);

let bloom = BloomFilterBuilder::default().with_bitmap_data(bitmap, FilterSize::Whatever);
```

Where the `max_key` value is derived from the `FilterSize`, so it is static across instances, so you don't need to serialise it to reconstruct the filter later (it wouldn't make much sense to mix differing filter sizes). It can be derived from the bitmap length!

This is also true for the `VecBitmap` too!

siennathesane replied:
~i can remove the documentation reference but since the other two impls use `serde`, i think it's best to leave it so the api is consistent.~

i read this wrong. yes, you are correct. i think leaving the note is prolly important for consumers since the example tests use `serde_json`. it took over a minute with `serde_json` instead of 150ms with `bincode`.

personally, i am not going to use serialization, but i think it's important for others who may have different needs

domodwyer replied:
Perhaps we can remove serde entirely from this type? Or implement serde as returning just the raw bytes rather than serialising the struct fields.

I can do this in a follow-up 🙏

siennathesane replied:
sure, i think letting you do it in a follow up would be better. i'm not an expert in bloom filters

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+        self.bitmap.freeze()
+    }
+
+    pub fn max_key(&self) -> usize {
+        self.max_key
+    }
+
+    pub fn from_bytes(bitmap: Bytes, max_key: usize) -> Self {
+        Self {
+            max_key,
+            bitmap: BytesMut::from(bitmap),
+        }
+    }
```
Currently the `max_key` property is not exposed in the public API; instead we expose the `FilterSize` enum and perform the mapping to a maximum key value internally instead. 

This hides this implementation detail from the user - they only ever talk about filter sizes. I think it would be valuable to keep the `max_key` out of the API. Fortunately the `max_key` can be inferred from the `bitmap.len()` by reversing this function:

https://github.com/domodwyer/bloom2/blob/e033136ae3e07429ef49489d18146156f08484e0/src/bloom.rs#L154-L156

So that then the constructor can forget about `max_key` entirely!

```rust
/// Docs here...
///
/// # Panics
///
/// If `bitmap` is incorrectly sized, this method panics.
pub fn from_bytes(bitmap: impl Into<Bytes>) -> Self {
    let max_key = len_to_filter_size(bitmap.len());
    // ...
}
```

This will keep `max_key` out of the public API and keep the interface simple. It'd be good to assert the correct bitmap length in `BloomFilterBuilder::with_bitmap_data()` to catch any mistakes though.

And as a bonus, the `impl Into<Bytes>` means we can accept `Vec<u8>` and friends too and handle the conversion automatically 🚀

siennathesane replied:
i think i updated it based on your intent? not 100% sure what you meant about "inverting" the method

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+        self.bitmap.resize(self.bitmap.len(), 0);
+    }
+
+    pub fn bitmap(self) -> Bytes {
```
I would perhaps call this `freeze()` too as it's specific to the `BytesBitmap`, and so the user is likely familiar with the freeze method on the `BytesMut`, so the semantics are clear to the user!

siennathesane replied:
done.

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+        (vec, self.max_key)
+    }
+
+    pub(crate) fn shrink(&mut self) {
+        self.bitmap.resize(self.bitmap.len(), 0);
+    }
```
I'm not sure this does anything - the docs for `resize()` speak in terms of buffer _length_ - not capacity. 

It seems this only grows or truncates the length of the buffer without modifying the allocated capacity, and so calling `resize(current_length)` will never do anything.

siennathesane replied:
yeah i wasn't 100% sure what to do here either. since there is no compression i guess it's really not necessary

domodwyer replied:
We definitely should implement compression! I can do it in a follow-up before a release though if you want :+1:

siennathesane replied:
that works for me!

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+}
+
+impl BytesBitmap {
+    pub(crate) fn into_parts(self) -> (Vec<usize>, usize) {
+        let mut vec = Vec::with_capacity(self.bitmap.len() / size_of::<usize>());
+        let chunks = self.bitmap.chunks_exact(size_of::<usize>());
+        for chunk in chunks {
+            vec.push(usize::from_ne_bytes(chunk.try_into().unwrap()));
+        }
+        (vec, self.max_key)
+    }
```
I don't think anything ever calls this, so that would mean this bitmap can't be compressed into the `CompressedBitmap` representation :fire:

We should keep it so fast read / write and space-inefficient bitmaps can be compressed into the fast read / space efficient forms.

siennathesane replied:
sure, i'll remove it. `vec` had it so i just ported it

domodwyer replied:
This is needed for compression - we just need to have a `compress()` method call it.

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+        let offset = index_for_key(key);
+        let byte_offset = offset * size_of::<usize>();
+
+        let slice = &mut self.bitmap[byte_offset..byte_offset + size_of::<usize>()];
+        let mut num = usize::from_ne_bytes(slice.try_into().unwrap());
+
+        if value {
+            num |= bitmask_for_key(key);
+        } else {
+            num &= !bitmask_for_key(key);
+        }
```
You could probably index the specific byte and use `get_mut()` instead of creating slices and converting into usize? The compiler might be smart enough to optimise this as-is though.

This whole type is a byte-wise representation (which meaningfully differs from the `VecBitmap` repr) so I think it's totally valid to optimise for that.

siennathesane replied:
i think this is one of those "mental complexity vs raw performance" challenges where i opted for "can someone else understand this later?". i think the biggest penalty here is the inevitable branch.

i did get curious tho, so i tested it with:

```
// old
if value {
    num |= bitmask_for_key(key);
} else {
    num &= !bitmask_for_key(key);
}

// new oneline
let bitmask = 1 << (key % (u64::BITS as usize));
num = (num & !bitmask) | (bitmask & (value as usize));
```

and this is what it looks like:
```
bytes_bitmap_insert_true
                        time:   [804.32 ps 806.40 ps 808.66 ps]
                        change: [-3.6809% -2.9364% -2.1772%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

bytes_bitmap_insert_false
                        time:   [831.44 ps 842.49 ps 855.27 ps]
                        change: [+1.9304% +2.9006% +4.0927%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  8 (8.00%) high mild
  2 (2.00%) high severe

bytes_bitmap_lookup_hit time:   [816.60 ps 822.18 ps 828.08 ps]
                        change: [+0.5063% +0.9796% +1.4763%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

bytes_bitmap_lookup_miss, different block
                        time:   [804.50 ps 806.31 ps 808.19 ps]
                        change: [-0.9178% -0.4502% +0.0686%] (p = 0.06 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

bytes_bitmap_lookup miss, same block
                        time:   [809.17 ps 814.55 ps 821.69 ps]
                        change: [-1.0797% -0.4035% +0.3524%] (p = 0.29 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe
```

i think this is a case where trying to be smarter than the compiler is 6 of one, half dozen of a another because of the inevitable branch. i'll leave it as-is since this is easier to reason about and the compiler can do what it does best

domodwyer replied:
The branchless variant is a neat trick! I was actually talking about the code above the conditional that slices the bytes and converts the bytes into a `usize` though - there's not really a restriction that requires us to use usizes in the `Bytes` encoding, and using `u8` instead has some compelling benefits in that the compiler / we can vectorise operations over 32 or more bytes at once, vs the `usize` casting approach which'll operate on 8 bytes at once (a kind of SWAR).

siennathesane replied:
sure, do you want me to swap it to `u8` from `usize`?

domodwyer replied:
No bother - I'll swap it over in a follow-up!

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+            .zip(other.bitmap.chunks_exact(size_of::<usize>()));
+
+        for (a_chunk, b_chunk) in chunks {
+            let a = usize::from_ne_bytes(a_chunk.try_into().unwrap());
+            let b = usize::from_ne_bytes(b_chunk.try_into().unwrap());
+            result.put_slice(&(a | b).to_ne_bytes());
```
Likewise this could definitely be a byte-wise merge without having to cast up to a usize - the compiler might even auto-vectorise the byte-wise merge for a chunky speed increase because there'd be no branching in the loop body :rocket:

siennathesane replied:
so i tried a few variations of this and i can't get it to consistently pass the tests.

for example:
```
// fails the tests
for i in 0..self.bitmap.len() {
    result.put_u8(self.bitmap[i] | other.bitmap[i]);
}

// won't compile due to no language support for `&[u8] | &[u8]`
for (a, b) in self.bitmap.iter().zip(other.bitmap.iter()) {
    result.put_u8(a | b);
}

// fails tests
result.extend_from_slice(&self.bitmap.iter()
    .zip(other.bitmap.iter())
    .map(|(a, b)| a | b)
    .collect::<BytesMut>());
```

i'm like 85% sure llvm is going to auto-vectorize this a bit because the iterators guaranteed to be ordered and the same size (due to the assertion). i'm also fairly positive the casts are `noop` because they're realistically just compiler hints. the only other thing i could really do would be to add SIMD intrinsics but those are inherently `unsafe`. i'm not sure this can be optimized further.

if you have other ideas i'm open to them.

domodwyer replied:
I had a look at this and is definitely isn't autovectorised, and I was unable to convince it to do so! 

For example, even manual unrolling the loop doesn't autovectorise:

```rust
const WIDE: usize = 16;
let mut buf = [0; WIDE];

let mut result = BytesMut::with_capacity(self.bitmap.len());
for (a, b) in self
    .bitmap
    .chunks_exact(WIDE)
    .zip(other.bitmap.chunks_exact(WIDE))
{
    buf[0] = a[0] | b[0];
    buf[1] = a[1] | b[1];
    buf[2] = a[2] | b[2];
    buf[3] = a[3] | b[3];
    buf[4] = a[4] | b[4];
    buf[5] = a[5] | b[5];
    buf[6] = a[6] | b[6];
    buf[7] = a[7] | b[7];
    buf[8] = a[8] | b[8];
    buf[9] = a[9] | b[9];
    buf[10] = a[10] | b[10];
    buf[11] = a[11] | b[11];
    buf[12] = a[12] | b[12];
    buf[13] = a[13] | b[13];
    buf[14] = a[14] | b[14];
    buf[15] = a[15] | b[15];

    result.put_slice(&buf);
}
```

But using `std::simd` on nightly really shows how trivially vectorisable it is:

```rust
const WIDE: usize = 16;

use std::simd::*;

let mut result = BytesMut::with_capacity(self.bitmap.len());
for (a, b) in self
    .bitmap
    .chunks_exact(WIDE)
    .zip(other.bitmap.chunks_exact(WIDE))
{
    let a = u8x16::from_slice(&a);
    let b = u8x16::from_slice(&b);

    result.put_slice(&(a | b).to_array());
}
```

This (unsurprisingly) does produce vectorised operations:

```asm
LBB9_7:
        ldr q0, [x23]
        ldr q1, [x22]
        orr.16b v0, v1, v0   ; <--- aarch64 vector registers
        str q0, [sp, #64]
        add x0, sp, #16
        add x1, sp, #64
        mov w2, #16
        bl <bytes::bytes_mut::BytesMut as bytes::buf::buf_mut::BufMut>::put_slice
        add x23, x23, #16
        add x22, x22, #16
        subs x21, x21, #1
        b.ne LBB9_7
```

🤷

I was hoping this would be a little tweak and have the compiler autovectorise it (ideally to 32-wide / 256 bit registers on platforms that have it - my laptop does not). 

I agree it's probably not worth the additional maintenance burdon of a `simd` feature flag for the gain unfortunately (on the plus side, it shows how nice `std::simd` could be!).

siennathesane replied:
yeah it might be worth a simd flag later for sure!

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,255 @@
+    }
+}
+
+// impl serde::Serialize for BytesBitmap {
```
What's the plan for this comment? Looks like you wound up deriving the serde impls in the end.

siennathesane replied:
whoops, sorry meant to delete that. i was trying to figure out why `serde_json` was so slow (it was serializing 67MiB of 0s)

domodwyer commented on src/bloom.rs:
```diff
@@ -59,12 +58,18 @@ where
 /// [SipHash]: https://131002.net/siphash/
-impl std::default::Default for BloomFilterBuilder<RandomState, CompressedBitmap> {
-    fn default() -> BloomFilterBuilder<RandomState, CompressedBitmap> {
+impl<B> std::default::Default for BloomFilterBuilder<RandomState, B>
```
This is a breaking change, but also a semantic change - it would mean a user no longer can just `default()` some working bloom filter config which is a nice API to expose - it also makes the `with_bitmap()` redundant. Can we keep this default impl as-is?

domodwyer commented on src/bloom.rs:
```diff
@@ -313,19 +318,34 @@ where
+{
+    /// Minimise the memory usage of this instance by shrinking the
+    /// underlying vectors, discarding their excess capacity.
+    pub fn shrink_to_fit(&mut self) {
```
If we wind up making this a no-op (see other comment) then we could add it to the `Bitmap` trait and avoid the specialisation for both the `CompressedBitmap` and the `BytesBitmap` (and the `VecBitmap` would benefit too, though it and the bytes impl are usually statically sized so there's _usually_ nothing extra to shrink).

domodwyer reviewed (commented):
Thanks for this @siennathesane! Sorry it took me a while to respond, it's that time of year - I really appreciate the PR 🚀 

I'm definitely keen on this! I think perhaps we should feature-gate it behind a `bytes` feature though (even if on-by-default) so that only users to want `bytes` have to pull it in (a bit like the `serde` support) :+1:

siennathesane commented:
Tbh I copied the `Vec<u8>` impl and then modified it for `bytes`. I'll work on it.

siennathesane commented:
fixed the conflict with a rebase.

domodwyer commented on src/bitmap/bytes.rs:
```diff
@@ -0,0 +1,164 @@
+    }
+    
+    fn or(&self, other: &Self) -> Self {
+        assert_eq!(self.bitmap.len(), other.bitmap.len());
```
It's probably best to assert that the inputs are multiples of the chunk size as the code does not attempt to process any remaining bytes, so this acts as a correctness assert:

```suggestion
        assert_eq!(self.bitmap.len(), other.bitmap.len());
        assert_eq!(self.bitmap.len() % size_of::<usize>(), 0);
```

(this always holds, but the assert hardens against future changes)

domodwyer commented on src/bloom.rs:
```diff
@@ -315,13 +316,17 @@ where
     H: BuildHasher,
 {
-    /// Minimise the memory usage of this instance by by shrinking the
+    /// Minimise the memory usage of this instance by shrinking the
```
Bye 👋

domodwyer reviewed (dismissed):
Hey! I think this is basically good to go now if you're happy with it? I am happy to polish the remaining bits up in a follow-on PR before a release unless you want to tackle them :+1:

Sorry for the delay over the holidays - I really needed a bit of a disconnect. I hope you had a good break too!

siennathesane commented:
i'm happy with it if you are! and no worries, gave me a chance to work on other things :)

domodwyer committed:
style: fmt

Run "cargo fmt".

domodwyer commented:
Thank you @siennathesane! Much appreciated :)

domodwyer committed:
test(bench): gate bytes impl behind cfg

domodwyer committed:
fix: move use for BytesBitmap into test mod

domodwyer committed:
refactor: use one cfg gate for bytes

Instead of cfg-ing each block, use one module-level feature cfg and a
conditional import - this avoids exposing the module path unless the
feature is enabled.

siennathesane commented:
When you finish with the fixes you wanted to make, would you mind publishing a new version?

domodwyer commented:
Absolutely! 

I'll work on some new benchmarks and other bits and cut a release soon - in the meantime you can [pin `bloom2` to the git hash](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#choice-of-commit) to pick up the `Bytes` work with something like:

```toml
bloom2 = { git = "https://github.com/domodwyer/bloom2.git", rev = "4154df9347fa0c28c3420635e6603e26e8a9a58c" }
```

Thanks again @siennathesane 🙏
